tracing-forest/src/tag.rs

```diff
@@ -1,267 +1,193 @@
-//!         RequestWarn,
-//!         #[tag(lvl = "error", msg = "request.error", macro = "request_error")]
-//!         RequestError,
-//!         #[tag(lvl = "trace", msg = "security.access", icon = '🔓', macro = "security_access")]
-//!         SecurityAccess,
-//!         #[tag(lvl = "info", msg = "security.info", icon = '🔒', macro = "security_info")]
-//!         SecurityInfo,
-//!         #[tag(lvl = "error", msg = "security.critical", icon = '🔐', macro = "security_critical")]
-//!         SecurityCritical,
-//!         #[tag(lvl = "trace", msg = "filter.trace", macro = "filter_trace")]
-//!         FilterTrace,
-//!         #[tag(lvl = "info", msg = "filter.info", macro = "filter_info")]
-//!         FilterInfo,
-//!         #[tag(lvl = "warn", msg = "filter.warn", macro = "filter_warn")]
-//!         FilterWarn,
-//!         #[tag(lvl = "error", msg = "filter.error", macro = "filter_error")]
-//!         FilterError,
-//!     }
-//! }
-//! ```
-use crate::{cfg_derive, cfg_json, fail};
-use tracing::Level;
+use crate::cfg_serde;
+use std::fmt;
+use tracing::{Event, Level};
 
-/// A type that can tag events with custom messages.
-///
-/// # Safety
+/// A type containing categorical information about where an event occurred.
 ///
-/// This trait is unsafe to implement as the method signatures are subject to
-/// change. Instead, [derive][tracing_forest_macros::Tag] it to ensure a correct
-/// implementation.
-///
-/// See [module level documentation][self] for how to use [`Tag`]s.
-pub unsafe trait Tag: 'static {
-    #[doc(hidden)]
-    fn as_field(&self) -> u64;
+/// See the [module-level documentation](crate::tag) for more details.
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub struct Tag {
+    /// Optional prefix for the tag message
+    prefix: Option<&'static str>,
+
+    /// Level specifying the important of the log.
+    ///
+    /// This value isn't necessarily "trace", "debug", "info", "warn", or "error",
+    /// and can be customized.
+    level: &'static str,
 
-    #[doc(hidden)]
-    fn from_field(value: u64) -> TagData;
+    /// An icon, typically emoji, that represents the tag.
+    icon: char,
 }
 
-pub(crate) type TagParser = fn(u64) -> TagData;
+impl Tag {
+    /// Returns a `Tag` constructed from an optional prefix and a custom level
+    /// and icon.
+    pub const fn new_custom_level(
+        prefix: Option<&'static str>,
+        level: &'static str,
+        icon: char,
+    ) -> Self {
+        Tag {
+            prefix,
+            level,
+            icon,
+        }
+    }
+
+    /// Returns a `Tag` constructed from an optional prefix and a level.
+    pub const fn new(prefix: Option<&'static str>, level: Level) -> Self {
+        match level {
+            Level::TRACE => Tag::new_custom_level(prefix, "trace", '📍'),
+            Level::DEBUG => Tag::new_custom_level(prefix, "debug", '🐛'),
+            Level::INFO => Tag::new_custom_level(prefix, "info", '💬'),
```

Comment:
I've noticed recently that since info is a super common level, it may be best to actually have no emoji here on these lines because it gets really noisy.

Reply:
I agree that the current emoji is pretty noisy, but I still kinda want to keep something there. Perhaps something less noisy like one of these?
```log
INFO     🪧 [info]: some information
INFO     🏷 [info]: some information
INFO     ◻️ [info]: some information
```
Currently leaning towards the last one.

Reply:
What's the last one? It's not rendering for me?

Reply:
It's just a white square lol. I figured it was very "vanilla" and that it wouldn't cause too much bloat if it appears a lot.

Reply:
Ahhh, yes but when you ahve black terminal and light colour text it adds up. So I think info should be two spaces instead.

Reply:
That ways it's really visual when something is warn/alert, because with the speech bubbles/white boxes it adds colour noise if that makes sense.